internal/credential/repository_store.go

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package credential
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/boundary/internal/db"
+	"github.com/hashicorp/boundary/internal/db/timestamp"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/pagination"
+	"github.com/hashicorp/boundary/internal/util"
+)
+
+// SubtypeStoreService defines the interface used to
+// list deleted store IDs of all subtypes.
+type SubtypeStoreService interface {
+	ListDeletedStoreIds(context.Context, time.Time, ...Option) ([]string, error)
```

Comment:
Deleted ID listing is still performed per-subtype as the deleted ID tables are per-subtype.

Reply:
I think we might want to get the deleted ids via a CTE that unions the deleted id tables. That would eliminate the need for this service and an implementation in each subtype, and since we don't need anything about the specific subtypes, just the ids, that seems like a simpler approach.